File: apps/scraper/query/sources/news.py
"""ASU News stories, searched by keyword."""

from __future__ import annotations

from scraper.core.types import QueryParam, QuerySource
from scraper.query.params import text, url


def to_url(params: dict[str, str]) -> str:
    """The news search results for a keyword."""
    return url("https://news.asu.edu/search", [("search", text(params, "keywords"))])


QUERY = QuerySource(
    key="news",
    description="Search ASU News for recent stories.",
    params=(QueryParam("keywords", "What the story is about.", required=True, example="robotics"),),
    to_url=to_url,
)

File: apps/engine/src/runtime/tools/knowledge/search/news.rs
//! Recent stories on ASU News.

use super::{LiveSource, Param};

/// The ASU News search.
pub struct News;

const PARAMS: &[Param] = &[Param::text(
    "keywords",
    "Topic to search for. Leave out for the newest stories.",
)
.example("robotics")];

impl LiveSource for News {
    fn key(&self) -> &'static str {
        "news"
    }

    fn hint(&self) -> &'static str {
        "ASU News stories"
    }

    fn label(&self) -> &'static str {
        "ASU News"
    }

    fn category(&self) -> &'static str {
        "news"
    }

    fn params(&self) -> &'static [Param] {
        PARAMS
    }
}
